packages/runtime/lib/runtime.js

```diff
@@ -0,0 +1,785 @@
+    const inspectorOptions = config.inspectorOptions
+    if (inspectorOptions) {
+      /* c8 ignore next 6 */
+      if (inspectorOptions.watchDisabled) {
+        logger.info('debugging flags were detected. hot reloading has been disabled')
+      }
+
+      inspector.open(inspectorOptions.port, inspectorOptions.host, inspectorOptions.breakFirstLine)
+    }
+
+    // Create all services, each in is own worker thread
+    for (const serviceConfig of config.services) {
+      // Setup forwarding of logs from the worker threads to the main thread
+      await this.#setupService(serviceConfig)
+    }
+
+    try {
+      // Make sure the list exists before computing the dependencies, otherwise some services might not be stopped
+      this.#servicesIds = config.services.map(service => service.id)
+
+      if (autoloadEnabled) {
+        checkDependencies(config.services)
+        this.#services = topologicalSort(this.#services, config)
+      }
+
+      // Recompute the list of services after sorting
+      this.#servicesIds = config.services.map(service => service.id)
+    } catch (e) {
+      await this.close()
+      throw e
+    }
+
+    this.#updateStatus('init')
+  }
+
+  async start () {
+    this.#updateStatus('starting')
+
+    // Important: do not use Promise.all here since it won't properly manage dependencies
+    for (const service of this.#servicesIds) {
+      await this.startService(service)
+    }
+
+    this.#updateStatus('started')
+
+    return this.#url
+  }
+
+  async stop () {
+    this.#updateStatus('stopping')
+
+    this.status = 'stopping'
+    await Promise.all(this.#servicesIds.map(service => this.stopService(service)))
+    this.status = false
+
+    this.#updateStatus('stopped')
+  }
+
+  async restart () {
+    this.emit('restarting')
+
+    await this.stop()
+    await this.start()
+
+    this.emit('restarted')
+
+    return this.#url
+  }
+
+  async close () {
+    this.#updateStatus('closing')
+    clearInterval(this.#metricsTimeout)
+    await this.stop()
+    this.#updateStatus('closed')
+  }
+
+  async startService (id) {
+    if (this.#servicesStarted[id]) {
+      throw new errors.ApplicationAlreadyStartedError()
+    }
+
+    // This is set here so that if the service fails while starting we track the status
+    this.#servicesStarted[id] = true
+
+    let service = await this.#getServiceById(id, false, false)
+
+    // The service was stopped, recreate the thread
+    if (!service) {
+      const config = this.#configManager.current
+      const serviceConfig = config.services.find(s => s.id === id)
+
+      await this.#setupService(serviceConfig)
+      service = await this.#getServiceById(id)
+    }
+
+    const serviceUrl = await service[kITC].send('start')
+
+    if (serviceUrl) {
+      this.#url = serviceUrl
+    }
+  }
+
+  async stopService (id) {
+    const service = await this.#getServiceById(id, false, false)
+
+    if (!service) {
+      return
+    }
+
+    this.#servicesStarted[id] = false
```

Comment:
What would happen if I call the `stopService` method while service is `starting`  (but not `started` yet). Is it possible a case when a service will miss the `stop` message ("stop" handler is not set for example)?